include/triton/backend/device_memory_tracker.h

```diff
@@ -0,0 +1,287 @@
+#include "triton/core/tritonserver.h"
+#include "triton/backend/backend_common.h"
+
+#ifdef TRITON_ENABLE_GPU
+#include <cupti.h>
+#endif
+
+static_assert(
+    sizeof(uint64_t) >= sizeof(uintptr_t),
+    "The implementation is storing address pointer as uint64_t, "
+    "must ensure the space for pointer is <= sizeof(uint64_t).");
+
+namespace triton { namespace backend {
+
+/// DeviceMemoryTracker is a backend utility provided to track the memory
+/// allocated for a particular model and associated model instance. This utility
+/// is often used for backend that set
+/// TRITONBACKEND_BackendAttributeSetEnableMemoryTracker to true, which provides
+/// additional information to Triton for making decision on model scaling and
+/// deployment.
+///
+/// Caveat: The memory tracker is implemented with CUPTI library which currently
+/// only supports single client/subscriber. This is an known limitation and as a
+/// result, the memory tracker can cause unexpected application failure if other
+/// component of the Triton process also uses CUPTI with a different
+/// configuration, for example, the framework used by the backend may have
+/// implemented similar profiler with CUPTI. Therefore, before enabling this
+/// memory tracker utilities, you should make sure that there is no other CUPTI
+/// client in the process. This tracker is implemented with the assumption that
+/// all other possible CUPTI clients are using the same implementation so that
+/// as long as all backends are compiled with this memory tracker and they may
+/// interact with CUPTI initialized externally to the backend without issues.
+///
+/// Typical usage:
+///
+/// The backend must implement TRITONBACKEND_GetBackendAttribute which will
+/// call TRITONBACKEND_BackendAttributeSetEnableMemoryTracker with proper
+/// arguments, and must implement TRITONBACKEND_ModelMemoryUsage.
+///
+/// On TRITONBACKEND_Initialize
+///  - Call DeviceMemoryTracker::Init
+///
+/// On TRITONBACKEND_GetBackendAttribute
+///  - Call TRITONBACKEND_BackendAttributeSetEnableMemoryTracker with value
+///    returned from DeviceMemoryTracker::Init
+///
+/// If true, call DeviceMemoryTracker::TrackThreadMemoryUsage and
+/// DeviceMemoryTracker::UntrackThreadMemoryUsage accordingly to track memory
+/// allocation in the scope between the function, the memory usage should be
+/// stored in MemoryUsage object that has the same lifetime as
+/// TRITONBACKEND_Model object.
+///
+/// On TRITONBACKEND_ModelMemoryUsage
+///  - Call MemoryUsage::SerializeToBufferAttributes to set the value
+///    to be returned.
+
+extern "C" {
+
+typedef struct TRITONBACKEND_CuptiTracker_t {
+  // C struct require extra implementation for dynamic array, for simplicity,
+  // the following assumptions are made to pre-allocate the array with max
+  // possible length:
+  //  - system / pinned memory allocation should only be on deviceId 0
+  //  - CUDA allocation will only be on visible CUDA devices
+  int64_t* system_byte_size_;
```

Comment:
naming here is a little confusing at first glance as it indicates a single "size", could we use something like:
```
system_memory_usage
cuda_memory_usage
```